src/devices/fakeMotionControl/fakeMotionControl.cpp

```diff
@@ -1360,134 +1399,359 @@ void FakeMotionControl::cleanup(void)
+            _ppids[j]=pid;
+        break;
+        case VOCAB_PIDTYPE_VELOCITY:
+            _vpids[j]=pid;
+        break;
+        case VOCAB_PIDTYPE_CURRENT:
+            _cpids[j]=pid;
+        break;
+        case VOCAB_PIDTYPE_TORQUE:
+           _tpids[j]=pid;
+        break;
+        default:
+        break;
+    }
     return true;
 }
 
-bool FakeMotionControl::setPidsRaw(const Pid *pids)
+bool FakeMotionControl::setPidsRaw(const PidControlTypeEnum& pidtype, const Pid *pids)
 {
     bool ret = true;
     for(int j=0; j< _njoints; j++)
     {
-        ret &= setPidRaw(j, pids[j]);
+        ret &= setPidRaw(pidtype, j, pids[j]);
     }
     return ret;
 }
 
-bool FakeMotionControl::setReferenceRaw(int j, double ref)
+bool FakeMotionControl::setPidReferenceRaw(const PidControlTypeEnum& pidtype, int j, double ref)
 {
-    int mode = 0;
-    getControlModeRaw(j, &mode);
-    if (mode != VOCAB_CM_POSITION_DIRECT &&
-        mode != VOCAB_CM_IDLE)
+    switch (pidtype)
     {
-        yError() << "setReferenceRaw: skipping command because  joint " << j << " is not in VOCAB_CM_POSITION_DIRECT mode";
+        case VOCAB_PIDTYPE_POSITION:
+            _ppids_ref[j]=ref;
+        break;
+        case VOCAB_PIDTYPE_VELOCITY:
+            _vpids_ref[j]=ref;
+        break;
+        case VOCAB_PIDTYPE_CURRENT:
+            _cpids_ref[j]=ref;
+        break;
+        case VOCAB_PIDTYPE_TORQUE:
+            _tpids_ref[j]=ref;
+        break;
+        default:
+        break;
     }
-
-   return false;
+    return true;
 }
 
-bool FakeMotionControl::setReferencesRaw(const double *refs)
+bool FakeMotionControl::setPidReferencesRaw(const PidControlTypeEnum& pidtype, const double *refs)
 {
     bool ret = true;
     for(int j=0, index=0; j< _njoints; j++, index++)
     {
-        ret &= setReferenceRaw(j, refs[index]);
+        ret &= setPidReferenceRaw(pidtype, j, refs[index]);
     }
     return ret;
 }
 
-bool FakeMotionControl::setErrorLimitRaw(int j, double limit)
+bool FakeMotionControl::setPidErrorLimitRaw(const PidControlTypeEnum& pidtype, int j, double limit)
 {
-    return NOT_YET_IMPLEMENTED("setErrorLimitRaw");
+    switch (pidtype)
+    {
+        case VOCAB_PIDTYPE_POSITION:
+            _ppids_lim[j]=limit;
+        break;
+        case VOCAB_PIDTYPE_VELOCITY:
+            _vpids_lim[j]=limit;
+        break;
+        case VOCAB_PIDTYPE_CURRENT:
+            _cpids_lim[j]=limit;
+        break;
+        case VOCAB_PIDTYPE_TORQUE:
+            _tpids_lim[j]=limit;
+        break;
+        default:
+        break;
+    }
+    return true;
 }
 
-bool FakeMotionControl::setErrorLimitsRaw(const double *limits)
+bool FakeMotionControl::setPidErrorLimitsRaw(const PidControlTypeEnum& pidtype, const double *limits)
 {
-    return NOT_YET_IMPLEMENTED("setErrorLimitsRaw");
+    bool ret = true;
+    for(int j=0, index=0; j< _njoints; j++, index++)
+    {
+        ret &= setPidErrorLimitRaw(pidtype, j, limits[index]);
```

Comment:
Using `&=` for a bool seemed a bit strange to me, because it is a bitwise and assignement. 
The proper operator should be `&&=` but apparently it does not exists in C++, and apparently `&=` on bool is perfectly safe, even if strange. [1]
[1] : https://stackoverflow.com/questions/2488406/why-doesnt-c-have-or-for-booleans